TACC/Core-CMS
# POC: For Style Plugin, Support Automatic Tag

## Overview

Intelligently choose HTML tag for user choice of class (for example, in Django Style).

## Tickets

* [FP-1318](https://jira.tacc.utexas.edu/browse/FP-1318), auto choose `<section>` tag for Sections created via:
    * https://github.com/TACC/Core-CMS/pull/430
    * https://github.com/TACC/Core-CMS/pull/431

## Changes

To our default `settings`:
* __Add `preferred_tag_for_class` template tag.__
* Add `DJANGOCMS_STYLE_TAGS_DEFAULT` (custom setting).
    * to allow dev to define default tag (should probably always be `div`)
    * to ensure the default tag is defined in only one location
* (Optional) Add `DJANGOCMS_STYLE_TAGS` ([docs](https://github.com/django-cms/djangocms-style#configuration)).
    * to support a use case of `DJANGOCMS_STYLE_TAGS_DEFAULT`

New files:
* (Optional) Override `djangocms-style` template.
    * to use `preferred_tag_for_class` template tag
* __Add `preferred_tag_for_class` template tag.__

## Screenshots

(pending)

## Testing

> local testing: [instructions](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes)
> remote testing: (not provided)

1. On any page, create a Style plugin.
1. Test _default_ tag choice logic:
    1. Choose TAG TYPE `div`, CLASS NAME `---------`.
    1. Confirm element added has tag of `<div>`.
1. Test _automatic_ tag choice logic:
    1. Choose TAG TYPE `div`, CLASS NAME `o-section o-section--style-light`.
    1. Confirm element added has tag of `<section>`.
    1. Choose TAG TYPE `div`, CLASS NAME `o-section o-section--style-dark`.
    1. Confirm element added has tag of `<section>`.
    1. Choose TAG TYPE `div`, CLASS NAME `c-callout`.
    1. Confirm element added has tag of `<aside>`.
    1. Choose TAG TYPE `div`, CLASS NAME `c-recognition c-recognition--style-light`.
    1. Confirm element added has tag of `<aside>`.
    1. Choose TAG TYPE `div`, CLASS NAME `c-recognition c-recognition--style-dark`.
    1. Confirm element added has tag of `<aside>`.
1. Test _manual_ tag choice logic:
    1. Choose TAG TYPE `article`, CLASS NAME `o-section o-section--style-light`.
    1. Confirm element added has tag of `<article>`.
    1. Choose TAG TYPE `article`, CLASS NAME `o-section o-section--style-dark`.
    1. Confirm element added has tag of `<article>`.
    1. Choose TAG TYPE `article`, CLASS NAME `c-callout`.
    1. Confirm element added has tag of `<article>`.
    1. Choose TAG TYPE `article`, CLASS NAME `c-recognition c-recognition--style-light`.
    1. Confirm element added has tag of `<article>`.
    1. Choose TAG TYPE `article`, CLASS NAME `c-recognition c-recognition--style-dark`.
    1. Confirm element added has tag of `<article>`.
